threads.go

```diff
@@ -0,0 +1,384 @@
+	if err != nil {
+		return
+	}
+
+	return lg.ID, coded, nil
+}
+
+// Put an existing record. See PutOption for more.
+func (t *threads) Put(ctx context.Context, rec thread.Record, opts ...tserv.PutOption) error {
+	// Get or create a log for the new rec
+	settings := tserv.PutOptions(opts...)
+	lg, err := t.getOrCreateLog(settings.Thread, settings.Log)
+	if err != nil {
+		return err
+	}
+
+	// Save the record locally
+	// Note: These get methods will return cached nodes.
+	block, err := rec.GetBlock(ctx, t.dagService)
+	if err != nil {
+		return err
+	}
+	event, ok := block.(*cbor.Event)
+	if !ok {
+		return fmt.Errorf("invalid event")
+	}
+	header, err := event.GetHeader(ctx, t.dagService, nil)
+	if err != nil {
+		return err
+	}
+	body, err := event.GetBody(ctx, t.dagService, nil)
+	if err != nil {
+		return err
+	}
+	err = t.dagService.AddMany(ctx, []format.Node{rec, event, header, body})
+	if err != nil {
+		return err
+	}
+
+	// Update head
+	t.SetHead(settings.Thread, lg.ID, rec.Cid())
+	return nil
+}
+
+// Pull for new records from the given log. See PullOption for more.
+// Local results are returned if the log is owned by this host.
+// Otherwise, records are pulled from the network at the log's addresses.
+func (t *threads) Pull(ctx context.Context, id thread.ID, lid peer.ID, offset cid.Cid, opts ...tserv.PullOption) ([]thread.Record, error) {
+	settings := tserv.PullOptions(opts...)
+
+	if t.PubKey(id, lid) == nil {
+		return nil, fmt.Errorf("could not find log")
+	}
+
+	if t.PrivKey(id, lid) != nil { // This is our own log
+		return t.pullLocal(ctx, id, lid, offset, settings.Limit)
+	}
+
+	// Pull from addresses
+	return t.service.pull(ctx, id, lid, offset, settings)
+}
+
+// Logs returns info about the logs in the given thread.
+func (t *threads) Logs(id thread.ID) []thread.LogInfo {
+	lgs := make([]thread.LogInfo, 0)
+	for _, lid := range t.ThreadInfo(id).Logs {
+		lg := t.LogInfo(id, lid)
+		lgs = append(lgs, lg)
+	}
+	return lgs
+}
+
+// Delete the given thread.
+func (t *threads) Delete(ctx context.Context, id thread.ID) error {
+	panic("implement me")
+}
+
+// getPrivKey returns the host's private key.
+func (t *threads) getPrivKey() ic.PrivKey {
+	return t.host.Peerstore().PrivKey(t.host.ID())
+}
+
+// createLog call util.CreateLog.
+func (t *threads) createLog() (info thread.LogInfo, err error) {
+	return util.CreateLog(t.host.ID())
+}
+
+// getOrCreateLog returns the log with the given thread and log id
+// If no log exists, a new one is created under the given thread.
+func (t *threads) getOrCreateLog(id thread.ID, lid peer.ID) (info thread.LogInfo, err error) {
+	info = t.LogInfo(id, lid)
+	if info.PubKey != nil {
+		return
+	}
+	info, err = t.createLog()
+	if err != nil {
+		return
+	}
+	err = t.AddLog(id, info)
+	return
+}
+
+// getOwnLoad returns the log owned by the host under the given thread.
+func (t *threads) getOwnLog(id thread.ID) (info thread.LogInfo) {
```

Comment:
Can I own more than one log in a thread?, if that makes sense it returns the first.

Reply:
You cannot - or in anycase if you did that would be unexpected.